src/browser/modules/D3Visualization/lib/visualization/components/Visualization.ts

```diff
@@ -17,216 +17,113 @@
+import {
+  nodeEventHandlers,
+  relationshipEventHandlers,
+  zoomEventHandler
+} from './mouseEventHandlers'
 import GraphStyle from 'browser/modules/D3Visualization/graphStyle'
 
 export type MeasureSizeFn = () => { width: number; height: number }
-export type VizObj = {
-  style: any
-  trigger: (_event: any, ..._args: any[]) => void
-  zoomInClick: (el: any) => { zoomInLimit: boolean; zoomOutLimit: boolean }
-  zoomOutClick: (el: any) => { zoomInLimit: boolean; zoomOutLimit: boolean }
-  boundingBox: () => void
-  resize: () => void
-  update: (options: {
-    updateNodes: boolean
-    updateRelationships: boolean
-    precompute?: boolean
-  }) => void
-}
 
-const noOp = () => undefined
-const vizFn = function (
-  el: SVGElement,
-  measureSize: MeasureSizeFn,
-  graph: Graph,
-  layout: Layout,
-  style: GraphStyle
-): VizObj {
-  const viz: VizObj = {
-    style,
-    // to be overwritten externally
-    trigger: noOp,
-    // to be defined later in this file, adding now for type safety
-    zoomInClick: () => ({
-      zoomInLimit: false,
-      zoomOutLimit: false
-    }),
-    zoomOutClick: () => ({
-      zoomInLimit: false,
-      zoomOutLimit: false
-    }),
-    boundingBox: noOp,
-    resize: noOp,
-    update: noOp
-  }
+export class Visualization<T extends Element = SVGElement> {
+  private root: Selection<T, unknown, BaseType, unknown>
+  private baseGroup: Selection<SVGGElement, unknown, BaseType, unknown>
+  private rect: Selection<SVGRectElement, unknown, BaseType, unknown>
+  private container: Selection<SVGGElement, unknown, BaseType, unknown>
+  private geometry: GraphGeometry
+  private zoomBehavior: ZoomBehavior<SVGGElement, unknown>
 
-  const root = d3Select(el)
-  const baseGroup = root.append('g').attr('transform', 'translate(0,0)')
-  const rect = baseGroup
-    .append('rect')
-    .style('fill', 'none')
-    .style('pointer-events', 'all')
-    // Make the rect cover the whole surface, center of the svg viewbox is in (0,0)
-    .attr('x', () => -Math.floor(measureSize().width / 2))
-    .attr('y', () => -Math.floor(measureSize().height / 2))
-    .attr('width', '100%')
-    .attr('height', '100%')
-    .attr('transform', 'scale(1)')
-
-  const container = baseGroup.append('g')
-  const geometry = new GraphGeometry(style)
+  public forceSim: ForceSimulation
 
   // This flags that a panning is ongoing and won't trigger
   // 'canvasClick' event when panning ends.
-  let draw = false
-
-  let updateViz = true
-  let isZoomClick = false
-
-  const onNodeClick = (_event: Event, node: VizNode) => {
-    updateViz = false
-    return viz.trigger('nodeClicked', node)
-  }
-
-  const onNodeDblClick = (_event: Event, node: VizNode) =>
-    viz.trigger('nodeDblClicked', node)
-
-  const onRelationshipClick = function (
-    event: Event,
-    relationship: Relationship
+  private draw = false
+  private isZoomClick = false
+
+  constructor(
+    element: T,
+    private measureSize: MeasureSizeFn,
```

Comment:
What does it mean to define a constructor argument as private? never seen this pattern befor

Reply:
This is the same as defining a private class property with the same name and then assigning it in the constructor: https://www.typescriptlang.org/docs/handbook/2/classes.html#parameter-properties